scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -2074,58 +2086,72 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
-      else
-        atPos(op, auto)(Term.ApplyUnary(op, simpleExpr(allowRepeated = true)))
+      else {
+        simpleExpr0(allowRepeated = true) match {
+          case Right(result) => atPos(op, auto)(Term.ApplyUnary(op, result))
+          case Left(error) =>
+            // maybe it is not unary operator but simply an ident `trait - {...}`
+            // we would fail here anyway, let's try to treat it as ident
+            simpleExprRest(op, canApply = true)
+        }
+      }
     }
 
   def simpleExpr(allowRepeated: Boolean): Term = autoPos {
+    simpleExpr0(allowRepeated) match {
+      case Right(result) => result
+      case Left(error) => throw error
+    }
+  }
+
+  private def simpleExpr0(allowRepeated: Boolean): Either[ParseException, Term] = autoPosE {
     var canApply = true
-    val t: Term = {
+    val t: Either[ParseException, Term] = {
       token match {
         case Literal() =>
-          literal()
+          Right(literal())
         case Interpolation.Id(_) =>
-          interpolateTerm()
+          Right(interpolateTerm())
         case Xml.Start() =>
-          xmlTerm()
+          Right(xmlTerm())
         case Ident(_) | KwThis() | KwSuper() | Unquote() =>
-          path() match {
+          Right(path() match {
             case q: Quasi => q.become[Term.Quasi]
             case path => path
-          }
+          })
         case Underscore() =>
           next()
-          atPos(in.prevTokenPos, in.prevTokenPos)(Term.Placeholder())
+          Right(atPos(in.prevTokenPos, in.prevTokenPos)(Term.Placeholder()))
         case LeftParen() =>
-          autoPos(inParensOrTupleOrUnit(location = NoStat, allowRepeated = allowRepeated))
+          Right(autoPos(inParensOrTupleOrUnit(location = NoStat, allowRepeated = allowRepeated)))
         case LeftBrace() =>
           canApply = false
-          blockExpr()
+          Right(blockExpr())
         case KwNew() =>
           canApply = false
           next()
-          atPos(in.prevTokenPos, auto) {
+          Right(atPos(in.prevTokenPos, auto) {
             template() match {
               case trivial @ Template(Nil, List(init), Self(Name.Anonymous(), None), Nil) =>
                 if (!token.prev.is[RightBrace]) Term.New(init)
                 else Term.NewAnonymous(trivial)
               case other =>
                 Term.NewAnonymous(other)
             }
-          }
+          })
         case MacroQuote() =>
-          macroQuote()
+          Right(macroQuote())
         case MacroSplice() =>
-          macroSplice()
+          Right(macroSplice())
         case MacroQuotedIdent(_) =>
-          macroQuotedIdent()
+          Right(macroQuotedIdent())
         case MacroSplicedIdent(_) =>
-          macroSplicedIdent()
+          Right(macroSplicedIdent())
         case _ =>
-          syntaxError(s"illegal start of simple expression", at = token)
+          Left(new ParseException(token.pos, "illegal start of simple expression"))
```

Comment:
since we are throwing it anyway, would it make sense to just throw it here? There would be no need for Either

Reply:
this either is used in 2 places, in one it is thrown, in second where I fix mentioned thing it is not thrown

Reply:
changed to `Try`as Albert suggested.